MHD & RMHD

ChrisZYJ opened the pull request:
## Description

* MHD
* HLLD Riemann Solver for MHD
* Powell's 8-wave method for MHD
* Relativistic MHD (RMHD)

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### Scope

- [x] This PR comprises a set of related changes with a common goal

## How Has This Been Tested?

- [x] HLL vs HLLD Dai and Woodward test (`examples/1D_dai_woodward` and `examples/1D_dai_woodward_hlld`)
Miyoshi, T. & Kusano, K. (2005). A multi-state HLL approximate Riemann solver for ideal magnetohydrodynamics.

![HLLvsHLLD_Dai_Woodward_test_Miyoshi_paper](https://github.com/user-attachments/assets/b3937628-efc4-49cb-9d38-c2d1abc9970d)

- [x] Orszag-Tang vortex (`examples/2D_orszag_tang`)
Orszag, S. A. & Tang, C.-M. (1979). Small-scale structure of two-dimensional magnetohydrodynamic turbulence.

https://github.com/user-attachments/assets/45206a28-972e-4034-83b7-c7f94f9964e5

- [x] RMHD Brio-Wu (`examples/1D_brio_wu_rmhd`)
Mignone, A. & Bodo, G. (2006). An HLLC Riemann solver for relativistic flows – II. Magnetohydrodynamics.

![RMHD_brio_wu_1 6k](https://github.com/user-attachments/assets/c0c0782d-1725-4307-a696-d71d3f638671)

- [x] RMHD Shock Cloud (`examples/2D_shock_cloud_rmhd`)
Mignone, A. & Bodo, G. (2006). An HLLC Riemann solver for relativistic flows – II. Magnetohydrodynamics.

https://github.com/user-attachments/assets/d8981619-5e90-4cb4-a991-332d2ddb4a06

### Performance

- [x] NCSA Delta A40

![compare_MHD_speed](https://github.com/user-attachments/assets/0ef800c4-91f7-4496-98ad-77da681810b8)

## Checklist

- [x] I have added comments for the new code
- [x] I added Doxygen docstrings to the new code
- [x] I have made corresponding changes to the documentation (`docs/`)
- [x] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [x] I have added example cases in `examples/` that demonstrate my new feature performing as expected.
They run to completion and demonstrate "interesting physics"
- [x] I ran `./mfc.sh format` before committing my code
- [x] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [x] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [x] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [x] Checked that the code compiles using NVHPC compilers
- [x] Checked that the code compiles using CRAY compilers
- [x] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [x] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [x] Enclosed the new feature via `nvtx` ranges so that they can be identified in profiles
- [ ] Ran a Nsight Systems profile using `./mfc.sh run XXXX --gpu -t simulation --nsys`, and have attached the output file (`.nsys-rep`) and plain text results to this PR
- [ ] Ran an Omniperf profile using `./mfc.sh run XXXX --gpu -t simulation --omniperf`, and have attached the output file and plain text results to this PR.
- [x] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

ChrisZYJ commented:
I think Codecov has mistakenly marked a lot of my code as uncovered, even though most/all of it is tested...

sbryngelson commented:
> I think Codecov has mistakenly marked a lot of my code as uncovered, even though most/all of it is tested...

Why is that?

ChrisZYJ commented:
I've double checked that the lines marked as uncovered by Codecov are indeed executed. When I commented them out and reran the tests locally, they all failed—confirming the lines are indeed being run. The Codecov logs also indicate that every test is being executed.

I think Codecov's issues are due to 2 main reasons: 

1. Multi-line code 

![image](https://github.com/user-attachments/assets/a15a8974-79bd-419b-bccd-5118a3919600)

2. fpp

![image](https://github.com/user-attachments/assets/1697f606-9d28-45a1-96c9-0ab2d3e24877)

While these factors don't always cause inaccurate coverage reports, they tend to cause more mistakes in complex subroutines like the Riemann solver.

sbryngelson commented:
> I've double checked that the lines marked as uncovered by Codecov are indeed executed. When I commented them out and reran the tests locally, they all failed—confirming the lines are indeed being run. The Codecov logs also indicate that every test is being executed.
> 
> I think Codecov's issues are due to 2 main reasons:
> 
> 1. Multi-line code
> 
> ![image](https://private-user-images.githubusercontent.com/120074479/427774217-a15a8974-79bd-419b-bccd-5118a3919600.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxMTU5ODUsIm5iZiI6MTc0MzExNTY4NSwicGF0aCI6Ii8xMjAwNzQ0NzkvNDI3Nzc0MjE3LWExNWE4OTc0LTc5YmQtNDE5Yi1iY2NkLTUxMThhMzkxOTYwMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyN1QyMjQ4MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iOTQwYjkxZDRhZWQ0NDJiZTE4NzFjNjI2YjA2OThiMjQ2MjE4ZWJhNDBiMTBmMDYzYzJmMjMwNjcyMjRkZDg2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FdlJyyIlhvB88R4scw4vf_Y2D2kGQoQAMPKrM4Nda3Y)
> 
> 2. fpp
> 
> ![image](https://private-user-images.githubusercontent.com/120074479/427774123-1697f606-9d28-45a1-96c9-0ab2d3e24877.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxMTU5ODUsIm5iZiI6MTc0MzExNTY4NSwicGF0aCI6Ii8xMjAwNzQ0NzkvNDI3Nzc0MTIzLTE2OTdmNjA2LTlkMjgtNDVhMS05NmM5LTBhYjJkM2UyNDg3Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyN1QyMjQ4MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZjFlMjQzYTVjN2M3NGE4ZjdhMmVjNWJkMWIzYWMzYjU0NGJlMjg5NjY4MzBmNmQzMGE2N2M1OTQ5NGEzOGNhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zXcSUayF339kDIiO-fL3z_OBmpdSs0goYB-evydoRIg)
> 
> While these factors don't always cause inaccurate coverage reports, they tend to cause more mistakes in complex subroutines like the Riemann solver.

Yes this makes sense. It certainly isn't perfect, and that's fine with me.

wilfonba reviewed (commented):
Great PR. Just two minor comments

ChrisZYJ commented on src/common/m_variables_conversion.fpp:
```diff
@@ -1053,8 +1158,6 @@ contains
         end do
         !$acc end parallel loop
 
-        !print *, 'I got here AA'
```
Ah I didn't add this. But I'll remove it!

wilfonba replied:
Oops. My bad. I was looking through this pretty quickly

ChrisZYJ commented:
@wilfonba Thanks so much for your suggestions! I've made the changes

sbryngelson commented on src/pre_process/include/2dHardcodedIC.fpp:
```diff
@@ -129,6 +130,34 @@
+        !   rho = 25/(36π)
+        !     p = 5/(12π)
+        !     v = (-sin(2π y), sin(2π x), 0)
+        !     B = (-sin(2π y)/sqrt(4π), sin(4π x)/sqrt(4π), 0)
```
can you remove the π character and replace with \pi

sbryngelson commented on src/pre_process/m_global_parameters.fpp:
```diff
@@ -392,6 +402,9 @@ contains
             patch_icpp(i)%qv = 0._wp
             patch_icpp(i)%qvp = 0._wp
             patch_icpp(i)%tau_e = 0._wp
+            patch_icpp(i)%Bx = dflt_real
```
How aout B(1), B(2), B(3) instead of three scalars? or alternatively B%x, B%y, etc. 

i would like to do this kind of thing across the whole code, and start using that standard for new PRs at least.

sbryngelson commented on src/simulation/m_global_parameters.fpp:
```diff
@@ -1142,19 +1176,24 @@ contains
+        Bxe = B_idx%end
 
-        !$acc update device(momxb, momxe, advxb, advxe, contxb, contxe, bubxb, bubxe, intxb, intxe, sys_size, buff_size, E_idx, alf_idx, n_idx, adv_n, adap_dt, pi_fac, strxb, strxe, chemxb, chemxe)
+        !$acc update device(momxb, momxe, advxb, advxe, contxb, contxe, bubxb, bubxe, intxb, intxe, sys_size, buff_size, E_idx, alf_idx, n_idx, adv_n, adap_dt, pi_fac, strxb, strxe, chemxb, chemxe, Bxb, Bxe)
```
I'm beginning to wonder if the new compilers really need the change from mom_idx to momxb, B_idx%beg to Bxb, etc. can you try taking this out? if it works, can you perform the more intrusive change of removing it in more places (checking that it isn't breaking things on nvidia or Frontier/amd.

sbryngelson commented on src/simulation/m_mhd.fpp:
```diff
@@ -0,0 +1,150 @@
+                        end do
+                    end if
+
+                    vx = q_prim_vf(momxb)%sf(k, l, q)
```
vx -> v%x or v(1)

sbryngelson commented on src/simulation/m_mhd.fpp:
```diff
@@ -0,0 +1,150 @@
+                    ! 2: vx  -> - (divB) * Bx
+                    ! 3: vy  -> - (divB) * By
+                    ! 4: vz  -> - (divB) * Bz
+                    ! 5: E   -> - (divB) * (v·B)
```
remove the dot. would like to keep the code all plain text no funky unicode

sbryngelson commented on src/simulation/m_riemann_solvers.fpp:
```diff
@@ -336,6 +352,16 @@ contains
+        real(wp) :: Ga_L, Ga_R ! Lorentz factor
+        real(wp) :: vdotB_L, vdotB_R, B2_L, B2_R
+        real(wp) :: b4x_L, b4y_L, b4z_L, b4x_R, b4y_R, b4z_R ! 4-magnetic field components
+        real(wp) :: mx_L, my_L, mz_L, mx_R, my_R, mz_R ! Conservative momentum variables
```
can we change all of the `mx_L` type variables to `mx%L`? or perhaps even `m%x%L`?
alternatively m(1,1) for (x,L), m(2,2) for (y,R), etc. same for all of the `_L and `_R`s more intrusively. i know this is a lot but really makes the code simpler.

sbryngelson reviewed (changes requested):
i left some requests for changes. overall very very good! as usual. exciting new stuff. can you update the README for the new capability and, if you have it, maybe we can even put an image in the readme as well.

sbryngelson commented:
@ChrisZYJ, I realized I've saddled you with a somewhat thankless task -- I hope it isn't so tough. I appreciate it very much, regardless. You can also look at the benchmarking logs to see how any of your code changes affected the `--case-optimization` feature.

ChrisZYJ commented:
@sbryngelson Thanks for the info! I think all tests should pass. I've updated my new features with the requested changes while keeping the code simple and readable:
- Removed unicode characters
- Grouped `%L` and `%R` (also `cm(i)%L`) using types
- Replaced `Bxb, Bxe` with `B_idx%beg, B_idx%end`

Apparently, removing `Bxb, Bxe` doesn't break AMD compilers. Points 2 and 3 could probably be applied throughout the code, but due to the the extent of changes, maybe it's best to do this in a separate PR so if anything goes wrong, we can pinpoint the error more easily.

sbryngelson commented:
> @sbryngelson Thanks for the info! I think all tests should pass. I've updated my new features with the requested changes while keeping the code simple and readable:
> 
> * Removed unicode characters
> * Grouped `%L` and `%R` (also `cm(i)%L`) using types
> * Replaced `Bxb, Bxe` with `B_idx%beg, B_idx%end`
> 
> Apparently, removing `Bxb, Bxe` doesn't break AMD compilers. Points 2 and 3 could probably be applied throughout the code, but due to the the extent of changes, maybe it's best to do this in a separate PR so if anything goes wrong, we can pinpoint the error more easily.

@ChrisZYJ I agree, we should separate concerns.

sbryngelson commented:
@ChrisZYJ This looks good but I want to wait until Phoenix is back online for those runners (a day or two, probably). Also I think you may have a case optimization problem? I can't remember why the Phoenix benchmark CI failed but if you download the logs you can check.

ChrisZYJ commented:
@sbryngelson You are right! Sorry that I kept thinking it was the runner's problem - really didn't expect it to work only on GPU without case-opt. In case anyone runs into this in the future: `var(i)%type` is not allowed for `!$acc private(...)`, but ``var%type(i)` is fine, and it only fails on GPU when case-optimization is enabled.

sbryngelson commented on examples/2D_axisym_shockwatercavity/case.py:
```diff
@@ -135,7 +135,7 @@
 SF = 400
 
-## making Nt divisible by SF
+## making Nt divisible By SF
```
remove all these changes to files from `by` to `By`

sbryngelson commented on src/common/m_variables_conversion.fpp:
```diff
@@ -1404,7 +1589,7 @@ contains
 
 #ifndef MFC_PRE_PROCESS
-    pure subroutine s_compute_speed_of_sound(pres, rho, gamma, pi_inf, H, adv, vel_sum, c_c, c)
+    subroutine s_compute_speed_of_sound(pres, rho, gamma, pi_inf, H, adv, vel_sum, c_c, c)
```
this might be important, why is this removed?

sbryngelson commented on src/simulation/m_riemann_solvers.fpp:
```diff
@@ -201,6 +204,19 @@ contains
                                        flux_gsrc_vf, &
                                        norm_dir, ix, iy, iz)
+        elseif (riemann_solver == 4) then
+            call s_hlld_riemann_solver(qL_prim_rsx_vf, qL_prim_rsy_vf, qL_prim_rsz_vf, dqL_prim_dx_vf, &
```
these really need to be refactored, they are all duplicating a ton of code...

sbryngelson commented on src/simulation/m_riemann_solvers.fpp:
```diff
@@ -2752,6 +2969,312 @@ contains
 
+    !> HLLD Riemann solver resolves 5 of the 7 waves of MHD equations:
+        !!      1 entropy wave, 2 Alfvén waves, 2 fast magnetosonic waves.
+    subroutine s_hlld_riemann_solver(qL_prim_rsx_vf, qL_prim_rsy_vf, qL_prim_rsz_vf, &
```
all of these riemann solvers duplicate a ton of code. please refactor

ChrisZYJ replied:
@sbryngelson Thanks for the comments. I've changed everything else, but I'm a bit hesitant to factor out the `subroutine ...` and the variable declarations using fpp because that breaks linting and file navigation. But I could do it if you wish.

sbryngelson replied:
i don't think we require fypp for this (though it might be useful in some parts). much of the Riemann solvers (hll/hllc/hlld) have the same code which should be pulled out into a separate subroutine (which would also be easier to read and add to).

ChrisZYJ replied:
I see what you mean now, and you're definitely right - a lot of code can be factored out between HLL and HLLC. But maybe this could be done in another PR? HLLD shares very little code with HLL or HLLC.

sbryngelson replied:
Understood, can you open a github issue for it and I'll merge this PR once all the runners are working?

ChrisZYJ replied:
Okay!
